### DISCUSSION TITLE
Testing a clearing-house structure with the IDS-Messaging-Service

### DISCUSSION THREAD

**Discussion by boxer-code** [International-Data-Spaces-Association/IDS-Messaging-Services#427] [Q&A] [answered]
Hey, 
i'm trying to test an existing clearing-house structure with connecting it to a connector or a testbed. At first i want to connect the core module of the IDS-Messaging-Service as a basis. I've added the maven artifact from step 1 and step 2 to the pom.xml from the connector. Now i'm struggeling with step 4 and 5. After i've added the annotation to the SpringBoot-Application in step 3, i'm confused where to add the Message-Handler from step 4 and step 5. Do these classes need to be in a special folder or in a special structure? And where do i have to add the function from step 5? Do i need to create a new class or could i add it to an existing class? 
And after adding these classes, is it enough to simply add the maven artifact of the clearing-house or do i have to add specific classes therefor as well? And how can i send a message to check wether my clearing-house setup is correct, when the set-up is complete? 
Thanks in Advance!

**Comment by ghost** [comment, marked as answer]
Hello and welcome!

The idea of the system behind it is actually quite simple:

**Receiving IDS-messages -> receiving a request and sending a response (Step 4):**
In order to receive and process incoming IDS-messages, in addition to an IDS-endpoint (/api/ids/data), so-called message-handlers are also required, which take care of processing of the incoming messages. The sample implementation shown in step 4 can handle incoming IDS-messages of type RequestMessage, see the SupportedMessageType annotation. The handleMessage method must be implemented, which executes certain logic when an IDS-message of this RequestMessage-type arrives. Once the logic has been processed, a response must be returned from the handleMessage for this message type, which is then sent back to the requesting Connector. Where exactly these handlers are located in the project and which further implementation structure is not relevant as long as the incoming IDS-message types all have corresponding handler classes. In summary, handlers take care of receiving specific types of IDS-messages and, if necessary, send back a response to the other connector.

**Sending of IDS-messages -> sending a request (Step 5):**
An IDS-message can be sent from any code location. The connector initiates a conversation with another connector by sending a (request) message and waiting for a response. 

An example of implementations of handlers and how to send IDS-messages has been implemented in the Dataspace Connector:
- Example of a handler:
https://github.com/International-Data-Spaces-Association/DataspaceConnector/blob/06ae90adf0d788e8812191e8745a6a3d09aecab5/src/main/java/io/dataspaceconnector/service/message/handler/type/DescriptionRequestHandler.java#L27 -> https://github.com/International-Data-Spaces-Association/DataspaceConnector/blob/06ae90adf0d788e8812191e8745a6a3d09aecab5/src/main/java/io/dataspaceconnector/service/message/handler/type/base/AbstractMessageHandler.java#L75
- Initiating a message to an IDS Metadata Broker and waiting for a response:
https://github.com/International-Data-Spaces-Association/DataspaceConnector/blob/06ae90adf0d788e8812191e8745a6a3d09aecab5/src/main/java/io/dataspaceconnector/service/message/GlobalMessageService.java#L82

For sending test messages to a ClearingHouse please contact the colleagues of the corresponding repo directly:
https://github.com/International-Data-Spaces-Association/ids-clearing-house-core
